Meter.c

```diff
@@ -266,89 +315,571 @@ static void BarMeterMode_draw(Meter* this, int x, int y, int w) {
 
 /* ---------- GraphMeterMode ---------- */
 
-#ifdef HAVE_LIBNCURSESW
-
-#define PIXPERROW_UTF8 4
-static const char* const GraphMeterMode_dotsUtf8[] = {
-   /*00*/" ", /*01*/"⢀", /*02*/"⢠", /*03*/"⢰", /*04*/ "⢸",
-   /*10*/"⡀", /*11*/"⣀", /*12*/"⣠", /*13*/"⣰", /*14*/ "⣸",
-   /*20*/"⡄", /*21*/"⣄", /*22*/"⣤", /*23*/"⣴", /*24*/ "⣼",
-   /*30*/"⡆", /*31*/"⣆", /*32*/"⣦", /*33*/"⣶", /*34*/ "⣾",
-   /*40*/"⡇", /*41*/"⣇", /*42*/"⣧", /*43*/"⣷", /*44*/ "⣿"
-};
+static void GraphMeterMode_createGraphData(Meter* this, unsigned int numRecords) {
+   assert(this->drawData == NULL);
+
+   GraphData* data = xCalloc(1, sizeof(GraphData));
+   data->numRecords = numRecords;
+
+   uint8_t maxItems = Meter_maxItems(this);
+
+   // recordNumCells determines number of cells needed to allocate for a graph
+   // data record. A record corresponds to a terminal column when drawn.
+   unsigned int recordNumCells;
+   if (GRAPH_HEIGHT >= 2 && maxItems <= 1) {
+      // 1 item only
+      // Index 0: log2 of scale
+      // Index 1: Number of dots total
+      recordNumCells = 2;
+   } else if (this->total > 0.0) {
+      // Percent graph
+      // Index 0..GRAPH_HEIGHT: Color cells
+      recordNumCells = GRAPH_HEIGHT;
+   } else {
+      // Non-percentage & dynamic scale
+      // Index 0: log2 of scale
+      // Index 1..(2*GRAPH_HEIGHT-1): Color cells
+      recordNumCells = 2 * GRAPH_HEIGHT;
+   }
+   data->recordNumCells = (uint16_t)recordNumCells;
+
+   GraphColorCell* cells = xCalloc((size_t)numRecords * recordNumCells,
+                                   sizeof(GraphColorCell));
+   if (GRAPH_HEIGHT < 2 || maxItems > 1) {
+      for (size_t i = 0; i < (size_t)numRecords * recordNumCells; i++) {
+         if (this->total > 0.0 || (i % recordNumCells) > 0) {
+            cells[i].cell.itemIndex = UINT8_MAX;
+         }
+      }
+   }
+   data->cells = cells;
+   this->drawData = data;
+}
 
-#endif
+static void GraphMeterMode_computeColors(Meter* this) {
```

Comment:
Excessive cyclomatic complexity. Please refactor to make this function more manageable …